SPIRE OIDC Federation to AWS v3

sanderson042 opened the pull request:
Kubernetes version of SPIRE OIDC Federation to AWS tutorial. This version makes most config decisions for the user user to make the tutorial experience easier for users new to SPIRE. This PR depends on Marcos Yacob's PR at https://github.com/spiffe/spire-tutorials/pull/13, which includes new YAML files for this tutorial.

evan2645 committed:
Remove Parent ID from node entry in k8s guide
Parent ID is not required when creating an entry via the CLI with the
`-node` flag.

Signed-off-by: Evan Gilman <evan@scytale.io>
Signed-off-by: Steve Anderson <steve.anderson@hpe.com>

Commit:
Draft of SPIRE OIDC federation to AWS. This v3 runs on Kubernetes instead of Linux.
Signed-off-by: Steve Anderson <steve.anderson@hpe.com>

Signed-off-by: Steve Anderson <steve.anderson@hpe.com>

Commit:
Delete some cruft.
Signed-off-by: Steve Anderson <steve.anderson@hpe.com>

Signed-off-by: Steve Anderson <steve.anderson@hpe.com>

Commit:
Fix broken table

Signed-off-by: Steve Anderson <steve.anderson@hpe.com>

Commit:
Add first set of corrections from Marcos Yacob.

Signed-off-by: Steve Anderson <steve.anderson@hpe.com>

Commit:
Correct and clarify DNS setup requirement

Signed-off-by: Steve Anderson <steve.anderson@hpe.com>

Commit:
Add more detail to the cleanup section.

Signed-off-by: Steve Anderson <steve.anderson@hpe.com>

steve-anderson-hpe committed:
No-op to test corect DCO config

Signed-off-by: Steve Anderson <steve.anderson@hpe.com>

steve-anderson-hpe committed:
Further edits based on problems found while testing.

Signed-off-by: Steve Anderson <steve.anderson@hpe.com>

steve-anderson-hpe committed:
Correct issues found by Marcos in his review.

Signed-off-by: Steve Anderson <steve.anderson@hpe.com>

steve-anderson-hpe committed:
Clarify/correct doc based on my own testing

Signed-off-by: Steve Anderson <steve.anderson@hpe.com>

steve-anderson-hpe committed:
More corrections from testing. Added troubleshooting section.

Signed-off-by: Steve Anderson <steve.anderson@hpe.com>

steve-anderson-hpe committed:
Read through once more and updated.

Signed-off-by: Steve Anderson <steve.anderson@hpe.com>

ajessup commented:
Super excited to see this @sanderson042. Will review this weekend and add any feedback.

ajessup commented on content/spire/try/oidc-federation-aws.md:
```diff
@@ -0,0 +1,415 @@
+---
+title: AWS OIDC Authentication
+description: Using OIDC to Authenticate SPIRE to AWS S3 on Kubernetes
```
This read a little strangely for me, since it wasn't quite clear what the subject of "on Kubernetes" was. Maybe "Using SPIRE and OIDC to authenticate workloads on Kubernetes to AWS S3"

sanderson042 replied:
Agreed, done.

ajessup committed:
Fixes minor type

Signed-off-by: Steve Anderson <steve.anderson@hpe.com>

ajessup commented on content/spire/try/oidc-federation-aws.md:
```diff
@@ -0,0 +1,415 @@
+* This tutorial cannot run on Minikube as the Kubernetes environment must be network accessible by AWS
+* You'll need the ability to configure a DNS A record for the SPIRE OIDC Discovery Provider document location (see next section)
+
+### Required DNS A Record for the OIDC Discovery Provider IP Address
```
I'm not sure why this section is here when it isn't possible to complete any of these steps until later, and you already call out that you need to complete the OIDC. Why do we ask the user to think about this now?

sanderson042 replied:
My thinking is that it's something the user will have to think about what tool to use to set up the A record, so I put it front and center here. Also, I didn't want to interrupt the flow of the k8s commands with this big explanation.
That said, I'm fine with moving it into the main flow. I think if I move the client deployment step to the testing section, this DNS step can be the last of part 1 or maybe a new part 2.

ajessup replied:
Yeah, your call, to me it's much more readable to just have a fairly short element in the pre-requisites that you'll need to be able to have a domain registered and the ability to set DNS A records associated with that domain, and move the rest into the main body.

sanderson042 replied:
Done. I moved the DNS steps inline into a new part 2 and included an nslookup verification step.

ajessup commented on content/spire/try/oidc-federation-aws.md:
```diff
@@ -0,0 +1,415 @@
+
+### Required DNS A Record for the OIDC Discovery Provider IP Address
+
+The SPIRE OIDC Discovery Provider provides a URL to the location of the discovery document specified by the OIDC protocol. After starting the spire-oidc service you'll need to configure a DNS A record to point to the external IP address of that service. In the YAML files set up for this tutorial, replace MY\_DISCOVERY\_DOMAIN with the FQDN of the planned DNS hostname. The FQDN value for MY\_DISCO ...
```
This paragraph confused me a little, and the statement "The FQDN value for MY\_DISCOVERY\_DOMAIN does not need to correspond to an existing domain on your network." confused me specifically as it implies the DNS entry doesn't need to be valid, but the entry does need to be valid. I need to have registered a domain name somewhere and have configured it's name-servers to have pointed to a DNS service that includes this domain.

Maybe a simpler way of expressing all of this would be:

"As part of this tutorial you will need to register a public DNS record that will resolve to the public IP address of your Kubernetes cluster. This will require you or an administrator to have registered a domain name (eg. `yutani.com`) with a domain name registrar, have configured it's name-server to point to a DNS service, and to be able to create an A record for a new subdomain (eg. `oidc-endpoint.yutani.com`) in that DNS service. If you don't have a registered domain name or access to a DNS service, services like [Google Domains](https://domains.google.com) can help you set one up for a fee.

sanderson042 replied:
Agreed, done.

ajessup commented on content/spire/try/oidc-federation-aws.md:
````diff
@@ -0,0 +1,415 @@
+```
+
+{{< info >}}
+As with any change to DNS, it will take minutes or hours for the new A record to propagate to DNS servers. This tutorial will not work until the A record is propagated. Once the A record is propagated, you should be able to view the discovery document at https://MY_DISCOVERY_DOMAIN/.well-known/openid-configuration .
````
I'm not sure this is true. Initial records are picked up quickly (since these are cache misses). But changes will require the TTL to expire, which by default is 3600 seconds (1 hour). This can be avoided by either (a) getting it right the first time, or (b) asking the user to specify a low TTL when creating the entry with their DNS provider.

ajessup commented on content/spire/try/oidc-federation-aws.md:
```diff
@@ -0,0 +1,415 @@
+
+Although this tutorial should contain all the info you need, the SPIRE OIDC Discovery Provider configuration file format is described on [GitHub](https://github.com/spiffe/spire/tree/master/support/oidc-discovery-provider).
+
+This tutorial uses Let's Encrypt to configure a certificate for the OIDC Discovery Provider domain. But other than editing the oidc-dp-configmap.yaml file as described in the previous section, no additional setup steps are required for Let's Encrypt.
```
Is a Lets Encrypt account a pre-requisite? If so it should be called out at the beginning in the pre-requisites section.

sanderson042 replied:
No Let's Encrypt account is needed per Eugene. Added "or account" to last sentence here. One could argue that this paragraph could be removed altogether since no user action is required.

ajessup commented on content/spire/try/oidc-federation-aws.md:
```diff
@@ -0,0 +1,415 @@
+| String | Description | Files to Change |
+| --- | --- | --- |
+| MY\_EMAIL\_ADDRESS | The terms of service for the Let's Encrypt certificate authority requires that you specify a valid email. Let's Encrypt is used to configure a certificate for the OIDC Discovery Domain (see the [Discovery Provider](#required-dns-a-record-for-the-oidc-discovery-provider-ip-address) info above). Example value: user@example.org | oidc-dp-configmap.yaml (1 instance) |
+| MY\_DISCOVERY\_DOMAIN | You must configure a public DNS A record for the IP address of the OIDC Discovery Provider. See the next section for details. Example value: `oidc-discovery.example.org` | ingress.yaml (2 instances), oidc-dp-configmap.yaml (1 instance), server-configmap.yaml (1 instance) |
```
This should focus on just what the user needs to do at this step - which is to decide on a custom sub domain name and to set it in the correct value in this file.

sanderson042 replied:
Agreed, done.

ajessup commented on content/spire/try/oidc-federation-aws.md:
```diff
@@ -0,0 +1,415 @@
+
+| String | Description | Files to Change |
+| --- | --- | --- |
+| MY\_EMAIL\_ADDRESS | The terms of service for the Let's Encrypt certificate authority requires that you specify a valid email. Let's Encrypt is used to configure a certificate for the OIDC Discovery Domain (see the [Discovery Provider](#required-dns-a-record-for-the-oidc-discovery-provider-ip-address) info above). Example value: user@example.org | oidc-dp-configmap.yaml (1 instance) |
```
It sounds like there's no setup required for lets encrypt - so this just needs to be a valid e-mail address that the user can receive e-mails to? If so, perhaps we just say something like that?

sanderson042 replied:
I think users may be curious about why they are putting an email address here. Will they get an email they have to respond to, or a newsletter, or ...? That said, I've tried to clarify and streamline this.

ajessup commented on content/spire/try/oidc-federation-aws.md:
````diff
@@ -0,0 +1,415 @@
+
+```console
+$ kubectl get pods spire-server-0 -n spire -o jsonpath='{.spec.containers[*].name}{"\n"}'
+```
````
What is the output one should expect to see here?

sanderson042 replied:
Added output as suggested.

ajessup commented on content/spire/try/oidc-federation-aws.md:
````diff
@@ -0,0 +1,415 @@
+    -f server-statefulset.yaml
+```
+
+To verify that the spire-server-0 pod has spire-server and spire-oidc containers, run:
````
While the spire-server pod is usually called `spire-server-0` there's no guarantee that it will be, and this may fail for some folks.  This version of the command will be more robust `kubectl get pods -n spire -l app=spire-server -o jsonpath='{.items[*].spec.containers[*].name}{"\n"}'`

sanderson042 replied:
Updated command as suggested.

ajessup commented on content/spire/try/oidc-federation-aws.md:
````diff
@@ -0,0 +1,415 @@
+    -f ingress.yaml 
+```
+
+## Verify the spire-server Stateful Set
````
Why is this section required here? Didn't we do it in the first tutorial? And hasn't it been tested again implicitly two steps ago?

sanderson042 replied:
Deleted "Verify the spire-server Stateful Set" as suggested.

ajessup commented on content/spire/try/oidc-federation-aws.md:
````diff
@@ -0,0 +1,415 @@
+Run the following command to verify that the EXTERNAL-IP value is present for the spire-oidc service. The spire-oidc Discovery Provider service must provide an external IP address for AWS to access the OIDC discovery document provided by spire-oidc.
+
+```console
+$ kubectl get service -n spire
````
More specific versions of this command that get the user what they need:

`kubectl get service -n spire spire-oidc`
(shows just the spire-oidc service)

or

`kubectl get service -n spire spire-oidc -o jsonpath={'.status.loadBalancer.ingress[*].ip'}`
(shows just the external IP address of the spire-oidc service)

sanderson042 replied:
Agreed. Changed doc to the first one.

ajessup commented on content/spire/try/oidc-federation-aws.md:
```diff
@@ -0,0 +1,415 @@
+
+At this point you can put the value of the spire-oidc EXTERNAL-IP in the required DNS A record, along with the value of the MY\_DISCOVERY\_DOMAIN domain that you specified in oidc-dp-configmap.yaml. For more details, see [Required DNS A Record for the OIDC Discovery Provider IP Address](#required-dns-a-record-for-the-oidc-discovery-provider-ip-address).
+
+## Deploy the Client
```
Why are we deploying this now, if we don't need it until later?

sanderson042 replied:
Moved client deployment to test section.

ajessup commented on content/spire/try/oidc-federation-aws.md:
```diff
@@ -0,0 +1,415 @@
+
+### Configure a DNS A Record to Point to the spire-oidc EXTERNAL-IP
+
+At this point you can put the value of the spire-oidc EXTERNAL-IP in the required DNS A record, along with the value of the MY\_DISCOVERY\_DOMAIN domain that you specified in oidc-dp-configmap.yaml. For more details, see [Required DNS A Record for the OIDC Discovery Provider IP Address](#required-dns-a-record-for-the-oidc-discovery-provider-ip-address).
```
Particularly since there are propagation delays, we should probably ask the user to verify their DNS entry at this point, using something like `nslookup` before proceeding

sanderson042 replied:
Done, covered in new part 2.

ajessup commented on content/spire/try/oidc-federation-aws.md:
```diff
@@ -0,0 +1,415 @@
+Note the following required accounts, prerequisites, and limitations before starting this tutorial:
+
+* You'll need access to the Kubernetes environment that you configured when going through [Kubernetes Quickstart](/spire/try/getting-started-k8s/) 
+* You'll need access to the AWS console to configure an identity provider, policy, role, and S3 bucket
```
More concretely, you'll need an account the AWS console that has the permissions to create new identity proivders, IAM policies, IAM roles and S3 buckets.

sanderson042 replied:
Done, clarified AWS permissions requirement.

ajessup commented on content/spire/try/oidc-federation-aws.md:
```diff
@@ -0,0 +1,415 @@
+         "Action": "sts:AssumeRoleWithWebIdentity",
+         "Condition": {
+           "StringEquals": {
+             "MY_DISCOVERY_DOMAIN:aud": "mys3",
```
This confused me since the text above says this should have been substituted for something. I'd suggest doing the substitution here and noting above something like "for example, if you used the domain `oidc-discovery.example.org` to host your OIDC discovery endpoint"

sanderson042 replied:
Agreed, done.

ajessup commented on content/spire/try/oidc-federation-aws.md:
````diff
@@ -0,0 +1,415 @@
+2. Start a shell on the client using the client pod name shown under **NAME** in the output of the last step.
+
+   ```console
+   $ kubectl exec -it client-74d4467b44-7nrs2 /bin/sh
````
You can simplify this and eliminate the previous step (on systems running a bash shell, at least, which will be most users) with the following command:

`kubectl exec -it $(kubectl get pods -o jsonpath={'.items[*]'.metadata.name}) /bin/sh`

sanderson042 replied:
Nice! I changed step 2 as suggested and moved the client deployment into step 1 instead of the existing step 1.

ajessup commented on content/spire/try/oidc-federation-aws.md:
````diff
@@ -0,0 +1,415 @@
+
+   ```console
+   # /opt/spire/bin/spire-agent api fetch jwt -audience mys3 -socketPath /run/spire/sockets/agent.sock | sed '2!d' | sed 's/[[:space:]]//g' > token
+   ```
````
There's a lot that can go wrong at this step. Consider asking the user to verify the contents of the token file to make sure it was written.

sanderson042 replied:
I've added a step to cat the token file.

ajessup commented on content/spire/try/oidc-federation-aws.md:
````diff
@@ -0,0 +1,415 @@
+   # /opt/spire/bin/spire-agent api fetch jwt -audience mys3 -socketPath /run/spire/sockets/agent.sock | sed '2!d' | sed 's/[[:space:]]//g' > token
+   ```
+
+4. Navigate to the AWS [Identity and Access Management (IAM) page](https://console.aws.amazon.com/iam/home#/home), logging in if necessary.
````
Consider just asking the user to make a note of the role ARN when they create the role in the first place.

sanderson042 replied:
Done, added a step to "Add the SPIFFE ID to the IAM Role" and removed a few here.

ajessup commented on content/spire/try/oidc-federation-aws.md:
````diff
@@ -0,0 +1,415 @@
+7. Run the following command, pasting the ARN in place of ROLE-NAME-ARN:
+
+   ```console
+   # AWS_ROLE_ARN=ROLE-NAME-ARN AWS_WEB_IDENTITY_TOKEN_FILE=token aws s3 cp s3://oidc-federation-test-bucket/test.txt test.txt
````
Consider making these two lines for readability. ie.
```
# AWS_ROLE_ARN=ROLE-NAME-ARN 
# AWS_WEB_IDENTITY_TOKEN_FILE=token aws s3 cp s3://oidc-federation-test-bucket/test.txt
```

ajessup replied:
Also I couldn't complete the exercise any further because I got the error "fatal error: Unable to locate credentials"

```
# echo $AWS_ROLE_ARN
arn:aws:iam::867136367285:role/oidc-federation-test-role
# echo $AWS_WEB_IDENTITY_TOKEN_FILE
token
cat $AWS_WEB_IDENTITY_TOKEN_FILE
eyJhbGciOiJFUzI1NiIsImtpZCI6Ik1PZ0NxU0loMElYcDlpWlBPbjNwWEhzQU5GcDlaY1E2IiwidHlwIjoiSldUIn0.eyJhdWQiOlsibXlzMyJdLCJleHAiOjE1ODU0Mjc4MjAsImlhdCI6MTU4NTQyNzUyMCwiaXNzIjoiaHR0cHM6Ly9vaWRjLmVsZWN0cmljLmRpcmVjdG9yeSIsInN1YiI6InNwaWZmZTovL2V4YW1wbGUub3JnL25zL2RlZmF1bHQvc2EvZGVmYXVsdCJ9.KUWZqeFinkfzVOImN7NeZ5SekPHh4281aEX5vQLZpQMoYaSRmTgOvzob_OZzbBb3Jm1yFvgslgYXXdKfmQl-Bw
```

sanderson042 replied:
I'm sorry it's not working for you. I did test it myself and got it working, though not after a snafu with a mismatched bucket name that was my error. 
The command seems to work with the environment variables on the same line as the command. Maybe it's easier for users to copy/paste one line from the doc than three?

ajessup replied:
I got this working once I created the bucket, and ran everything on the same line rather than creating the environment variables on separate lines.

ajessup reviewed (changes requested):
Improving, but I still wasn't able to actually complete the tutorial, there may be some steps missing? Some other fixes included or suggested.

sanderson042 reviewed (commented):
Sorry, I overlooked this PR and now the text has moved on. Could you please close?

I've fixed the S3 typos. I converted all the parts to numerical numbering rather than spelled out, since we're up to four parts now (e.g. " Part 3: Configure AWS Components" vs. " Part Three: Configure AWS Components").

steve-anderson-hpe committed:
Made all changes requested by Andrew Jessup on 28-March-2020.

The major change is to move the DNS info into the main flow of the
document as a new part 2.

Signed-off-by: Steve Anderson <steve.anderson@hpe.com>

ajessup commented:
Hi Steve - if you're addressing feedback you can just push the changes to the same branch that you created this PR from and it will be updated, there's no need to close this PR and re-open another one. 

If you've already opened another PR though, please point me to that and close this one.

sanderson042 commented:
My bad, I used the wrong term. I was talking about the brief commit you made for S3 typos: 69ffc24, not a PR. So closing it is not applicable.

As for this PR #125, yes, I'll keep it as is and push changes here. My major update is ready. I'm just working through a git/github issue.

sanderson042 commented:
Ok, AJ's suggestions are now incorporated. The doc is ready for another review as of commit 06e4209 (or technically the merge at 6b85d39).

ajessup committed:
Consolidates the pre-requisites to make it clearer

Signed-off-by: Steve Anderson <steve.anderson@hpe.com>

ajessup commented on content/spire/try/oidc-federation-aws.md:
```diff
@@ -0,0 +1,449 @@
+
+3. Click **Create bucket**.
+
+4. Under **Bucket name** type `oidc-federation-test-bucket`.
```
This won't work, because S3 buckets must have globaly unqiue names (that is, no two users anywhere can have the same AWS S3 bucket name). You'll have to ask them to make up a name and use that throughout the tutorial.

sanderson042 replied:
Oh wow, good catch. I am moving this S3 bucket creation section to the first section under Part 3 so it's more natural to have users substitute their actual S3 bucket name in the JSON blob later on. I'm also explaining this requirement here during bucket creation.

ajessup committed:
Minor fixes

Signed-off-by: Steve Anderson <steve.anderson@hpe.com>

ajessup commented:
Hooray! I was able to complete the the tutorial! Modulo the one remaining fix for unique s3 bucket names we should be good to go! It's looking great. Takes a very complex experience and makes it really clear and straightforward.

steve-anderson-hpe committed:
Address Andrew Jessup's latest review comments (S3 bucket names)

S3 bucket names must be globally unique, so:
--Moved "Create an AWS S3 Bucket and Test File" to start of AWS section
--Added info about globally unique S3 bucket name requirement
--Added MY_TEST_BUCKET S3 bucket name placeholder throughout

Also includes other minor miscellaneous edits for clarity.

Signed-off-by: Steve Anderson <steve.anderson@scytale.io>
Signed-off-by: Steve Anderson <steve.anderson@hpe.com>

sanderson042 commented:
Great, I'm glad your latest test was successful. Commit e0e3898 addresses the S3 bucket name issue. Thanks to you for your on point suggestions for improvements and thanks to Marcos for his YAML work.

mchurichi commented on content/spire/try/oidc-federation-aws.md:
```diff
@@ -0,0 +1,448 @@
+
+## Download Kubernetes YAML Files for this Tutorial
+
+To get the YAML files required for this tutorial, clone https://github.com/spiffe/spire-tutorials. The files for this tutorial are in the **k8s/oidc-aws** directory.
```
is there any rule of thumb to use `code` formatting? I'm used to read documentation that uses it when referencing commands, urls and file paths

sanderson042 replied:
Yeah, I agree we should use a monospace font for commands, URLs, file names, and paths. However, I'd prefer to leave live (clickable) URLs in the default bold+colored rendering automatically applied by Markdown/Hugo (e.g. https://jwt.io). The k8s names I've left in bold since the Quickstart for Kubernetes doc uses that convention, but eventually I think I'd like to use monospace for k8s names, too.

I've updated the doc and will push a commit soon.

sanderson042 commented on content/spire/try/oidc-federation-aws.md:
````diff
@@ -0,0 +1,448 @@
+```console
+$ kubectl get pods -n spire -l app=spire-server -o \
+    jsonpath='{.items[*].spec.containers[*].name}{"\n"}'
+spire-server spire-oidc
````
Sure, I'll split up the output here.

steve-anderson-hpe committed:
Fixes suggested by @mchurichi: font for paths and...

Clarify output for a certain k8s command

Signed-off-by: Steve Anderson <steve.anderson@scytale.io>
Signed-off-by: Steve Anderson <steve.anderson@hpe.com>

ajessup reviewed (approved):
LGTM!
